Friends-app

xT0PERx opened the pull request:
### Friends-app

[Demo](https://xt0perx.github.io/friends-app/)
[Code base](https://github.com/xT0PERx/friends-app)

The code is submitted in a dedicated feature branch.
Please, review.

zonzujiro commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,177 @@
+      }
+    })
+    .then((data) => {
+      saveUsers(data.results);
+      createUsersCards(ALL_USERS_CARDS);
```
Single responsibility - your functions should do only one job. As an example function in which you fetch users should only fetch them and not render, transform or process them in other ways. All these things should be done in another place, outside your function

xT0PERx replied:
Could you explain in more detail how I can process data without using functions in __fetch__. Perhaps I should save the data to a variable and then work with it, is this considered the best way?

zonzujiro replied:
No, the global variable not the best way. Your get function should return the result of fetch, which is a promise with needed data and then you handle this promise as usual.

zonzujiro reviewed (changes requested):
OK , I see common mistakes.  Please, go thru this check list:  

- Single responsibility - your functions should do only one job. As an example function in which you fetch users should only fetch them and not render, transform or process them in other ways. All this things should be done in another place, outside your function. Same applied to sort function, which usually do all types of sorting ;)
- You should handle fetch response status - https://www.tjvantoll.com/2015/09/13/fetch-and-errors/
- Don’t use names as `arr`, `array`, `data` or etc. They to generic. Don’t use things like `sortingByABC` and `sortingByZYX` - it’s unclear and just weird. `friendsArr` also bad idea - you show data type in variable name and in JS it’s redundant
- Prefer template strings to “createElement” API
- Your template strings which contains should have intendations as in HTML
- When you need to decide which sort condition were selected, prefer `value` of input to “check class name of pressed element”, or “compare pressed element to saved element” etc. You can have value “female”, get value and then just use it
- Don’t overuse global variables, keep data which you are using closest as possible to place where you using it. As an example, request url can be inside function which sends request

xT0PERx commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,177 @@
+
+  users.forEach(function (el) {
+    const card = document.createElement("li");
+    card.classList.add("userCard");
+    card.innerHTML = `<span class="name">${el.name.first} ${el.name.last}</span>
+    <img src="${el.picture.large}">
+    <span class="age">Age: ${el.dob.age}</span>
+    <span class="location">Location: ${el.location.city}</span>`;
```
Do I understand this point of the comment correctly? 
___Prefer template strings to “createElement” API___
This means that it is better for me to create elements not through a piece of  HTML code, but using a command such as __document.createElement()__? and then __innerHTML="bla-bla"__?

zonzujiro replied:
No, vice-versa. Template strings should be used

xT0PERx commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,177 @@
+const ALL_USERS_CARDS = allFriends.currentList;
+const USERS_LIST = document.getElementById("usersList");
+const FILTERS_MENU = document.getElementById("filters_menu");
+const ABC_BUTTON_FILTER = document.getElementById("Abc");
+const ZYX_BUTTON_FILTER = document.getElementById("Zyx");
+const MALE_BUTTON_FILTER = document.getElementById("male");
+const FEMALE_BUTTON_FILTER = document.getElementById("female");
+const YOUNGER_FILTER_BUTTON = document.getElementById("younger-ftr");
+const SENIOR_FILTER_BUTTON = document.getElementById("senior-ftr");
+const RESET_BUTTON = document.getElementById("reset");
```
Do I understand this point of the comment correctly?
___When you need to decide which sort condition were selected, prefer value of input to “check class name of pressed element”, or “compare pressed element to saved element” etc. You can have value “female”, get value and then just use it___
It is better to assign a __value__ to the __button__ element and use it in my code than to assign to the element __id__?

zonzujiro replied:
> value of input

vs 

> value to the button

:)

And WDYM under?

>  to assign to the element id

`value` will allow you to understand what element were clicked

zonzujiro replied:
> value of input

vs 

> value to the button

:)

And WDYM under?

>  to assign to the element id

`value` will allow you to understand what element was clicked

xT0PERx replied:
To understand which element was clicked, I use  __\<button id="name-ascend"__. My question is as follows. For elements which, according to the specification, provide such properties as __"value"__ it is better to use them, and for elements whitch dont have __"value"__  - i use __id__(if I want to apply to them JS code). 
I.e __\<button value="name-ascend"___ instead __\<button id="name-ascend"__. 
I seem to have caught up, but I still want to hear an expert's opinion :)

zonzujiro replied:
> My question is as follows. For elements which, according to the specification, provide such properties as "value" it is better to use them

Yes.

> and for elements whitch dont have "value" - i use id(if I want to apply to them JS code).

No. `id` does not replace `value`. It is just a unique identificator and it can be used here and there. Generally speaking, it has it's own "do's and don'ts".

You don't need `button`, you need `radio` because you have two buttons and both buttons can have two states - checked and unchecked. It is a list of radio buttons.

So what you should do is replace `button` with `input type='radio'` and then get the value of input from the `change` event.

zonzujiro commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,177 @@
+  createUsersCards(decreaseSearch);
+}
+
+function sortAscend() {
+  const youngers = ALL_USERS_CARDS_FILTERS.sort(
+    (a, b) => a.dob.age - b.dob.age
+  );
+  resetCards();
+  createUsersCards(youngers);
+}
+
+function descendSort() {
+  const seniors = ALL_USERS_CARDS_FILTERS.sort((a, b) => b.dob.age - a.dob.age);
+  resetCards();
+  createUsersCards(seniors);
+}
```
Same functions. You can use one instead. Only order of arguments matters

zonzujiro commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,177 @@
+  createUsersCards(seniors);
+}
+
+function sortByMale() {
+  const filterMale = ALL_USERS_CARDS_FILTERS.filter(
+    (elem) => elem.gender === "male"
+  );
+  resetCards();
+  createUsersCards(filterMale);
+}
+
+function sortByFemale() {
+  const filterFemale = ALL_USERS_CARDS_FILTERS.filter(
+    (elem) => elem.gender === "female"
+  );
+  resetCards();
+  createUsersCards(filterFemale);
+}
```
Same with same functions

zonzujiro commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,177 @@
+  USERS_LIST.innerHTML = "";
+}
+
+// function sortWithAppliedFilters() {}
```
Please remove unused code from PR

zonzujiro commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,177 @@
+  }
+});
+
+function sortByAscendName() {
+  const increaseSearch = ALL_USERS_CARDS_FILTERS.sort((a, b) => {
+    if (a.name.first > b.name.first) {
+      return 1;
+    }
+    if (a.name.first < b.name.first) {
+      return -1;
+    }
+    return 0;
+  });
+
+  resetCards();
+  createUsersCards(increaseSearch);
+}
+
+function sortByDescendName() {
+  const decreaseSearch = ALL_USERS_CARDS_FILTERS.sort((a, b) => {
+    if (a.name.first < b.name.first) {
+      return 1;
+    }
+    if (a.name.first > b.name.first) {
+      return -1;
+    }
+    return 0;
+  });
+  resetCards();
+  createUsersCards(decreaseSearch);
+}
```
Same functions

zonzujiro commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,177 @@
+  USERS_LIST.appendChild(lists);
+}
+
+FILTERS_MENU.addEventListener("click", function ({ target }) {
+  switch (target) {
+    case NAME_ASCEND_BUTTON_FILTER:
+      sortByAscendName();
+      break;
+    case NAME_DESCEND_BUTTON_FILTER:
+      sortByDescendName();
+      break;
+    case YOUNGER_FILTER_BUTTON:
+      sortAscend();
+      break;
+    case SENIOR_FILTER_BUTTON:
+      descendSort();
+      break;
+    case MALE_BUTTON_FILTER:
+      sortByMale();
+      break;
+    case FEMALE_BUTTON_FILTER:
+      sortByFemale();
+      break;
+    case RESET_BUTTON:
+      resetFilters();
+      break;
+  }
+});
```
When you need to decide which sort of condition was selected, prefer value of input instead “check classname of pressed element”, or “compare pressed element to saved element” etc. You can have value female, get value and then just use it

zonzujiro commented:
Please, resolve outdated :)

zonzujiro commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,175 @@
+  createUsersCards(ageSort);
+}
+
+function sortByGender() {
```
```jsx
function sortByGender(gender) {
```

And then just

```jsx
    return ALL_USERS_CARDS_FILTERS.filter(
      (elem) => elem.gender === gender
    );
```

zonzujiro commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,175 @@
+  USERS_LIST.innerHTML = "";
+}
+
+function offChecked() {
```
`resetInputs`

xT0PERx commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,175 @@
+  users.forEach(function (el) {
+    const card = document.createElement("li");
+    card.classList.add("userCard");
+    template = `<span class="name">${el.name.first} ${el.name.last}</span>
```
No, I haven't worked with Python. It's all about my carelessness - I'll fix it.

xT0PERx commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,175 @@
+     <img src="${el.picture.large}">
+     <span class="age">Age: ${el.dob.age}</span>
+     <span class="location">Location: ${el.location.city}</span>`;
+    card.insertAdjacentHTML("beforeEnd", template);
```
I read in the specification that this method works faster than **innerHtml**, I decided to use it.

xT0PERx commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,175 @@
+}
+
+FILTERS_MENU.addEventListener("click", function ({ target }) {
+  if (target.nodeName === "INPUT") {
```
Since my __input__ with the __label__ is considered __one target__, nothing else comes to my mind than this check. I would be glad to your advice. Thank.

zonzujiro commented:
Resolve outdated :)

xT0PERx commented:
I corrected everything except the check __if (target.nodeName === "INPUT")__ - but unfortunately I don't see another way out - I would be grateful for advice on this matter. If I suddenly missed something else - I apologize - but reviewed it several times - in my opinion, I corrected all the previous comments.

zonzujiro commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,158 @@
+function search(people) {
+  const search = searchInput.value.toLowerCase();
+
+  const peopleSearch = people.filter(
```
Just return, without variable.

zonzujiro commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,158 @@
+  createUsersCards(featuredUsers);
+});
+
+function search(people) {
```
Verb

zonzujiro commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,137 @@
+    return;
+  }
+
+  if (target.name === "search") {
```
Can be a good idea to separate form into two fieldsets.

One with sorters and another with a search filter. Then, you will don't need this check. So this event listener will don't know anything about the existence of search input.

zonzujiro replied:
And about the reset button also ;)

zonzujiro commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  const sortedCases = {
+    sort_by_alphabet: (users) => nameSorters[target.value](users),
```
I don't like `snake_cases`. In JS you should use `camelCase`. So you can rename `name` :)

zonzujiro commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,137 @@
+
+const ageSorters = {
+  younger: (users) => users.sort(compareAge),
+  senior: (users) => users.sort((a, b) => compareAge(b, a)),
```
`older` :)

zonzujiro commented on submissions/xT0PERx/Friends app/js/index.js:
```diff
@@ -0,0 +1,137 @@
+  createUsersCards(featuredUsers);
+});
+
+function seekNames(people) {
```
`findByName`?

zonzujiro reviewed (approved):
LGTM.

Comments just FYI
